### ISSUE TITLE
user defined functions in Parameter asteval expressions overwritten by __deepcopy__

### ISSUE THREAD

**Issue by zobristnicholas** [lmfit/lmfit-py#507]
Hi,
Using a custom asteval.Interpreter class with user defined functions to initialize a Parameters object does not work. It looks like this is unintended behavior.

The Parameters object is created correctly, but it is copied before the fit and the custom function is not transferred. It looks like the just the __deepcopy__ method for the Parameters object needs to be updated to address this.

Here is a minimal working example:
```
import numpy as np
import lmfit as lm
import asteval

def myfun(x):
    return x**3

params = lm.Parameters(asteval=asteval.Interpreter(usersyms={"myfun": myfun}))
params.add("a", value=2)
params.add("b",` expr="myfun(a)")

print(params)

xx = np.linspace(0, 1, 10)
yy = 3 * xx

model = lm.Model(lambda x, a: a * x)
result = model.fit(yy, params=params, x=xx)
```

My system information is:
Python: 3.6.5 |Anaconda, Inc.| (default, Apr 26 2018, 08:42:37)
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]

lmfit: 0.9.11, scipy: 1.1.0, numpy: 1.14.3, asteval: 0.9.13, uncertainties: 3.0.2, six: 1.11.0

The result is:
```
Parameters([('a', <Parameter 'a', 2, bounds=[-inf:inf]>), ('b', <Parameter 'b', 8, bounds=[-inf:inf], expr='myfun(a)'>)])
NameError
   <_ast.Module object at 0x11444a518>
name 'myfun' is not defined
Traceback (most recent call last):
  File "test.py", line 18, in <module>
    result = model.fit(yy, params=params, x=xx)
  File "/Users/nicholaszobrist/anaconda3/lib/python3.6/site-packages/lmfit/model.py", line 803, in fit
    params = deepcopy(params)
  File "/Users/nicholaszobrist/anaconda3/lib/python3.6/copy.py", line 161, in deepcopy
    y = copier(memo)
  File "/Users/nicholaszobrist/anaconda3/lib/python3.6/site-packages/lmfit/parameter.py", line 124, in __deepcopy__
    _pars.add_many(*parameter_list)
  File "/Users/nicholaszobrist/anaconda3/lib/python3.6/site-packages/lmfit/parameter.py", line 364, in add_many
    self.__setitem__(para.name, para)
  File "/Users/nicholaszobrist/anaconda3/lib/python3.6/site-packages/lmfit/parameter.py", line 138, in __setitem__
    self._asteval.symtable[key] = par.value
  File "/Users/nicholaszobrist/anaconda3/lib/python3.6/site-packages/lmfit/parameter.py", line 791, in value
    return self._getval()
  File "/Users/nicholaszobrist/anaconda3/lib/python3.6/site-packages/lmfit/parameter.py", line 773, in _getval
    check_ast_errors(self._expr_eval)
  File "/Users/nicholaszobrist/anaconda3/lib/python3.6/site-packages/lmfit/parameter.py", line 21, in check_ast_errors
    expr_eval.raise_exception(None)
  File "/Users/nicholaszobrist/anaconda3/lib/python3.6/site-packages/asteval/asteval.py", line 261, in raise_exception
    raise exc(self.error_msg)
NameError: at expr='<_ast.Module object at 0x11444a518>'
```

**Comment by newville** [comment]
@zobristnicholas   It looks to me like this is not a copy/deepcopy problem but a question of "what is the best way to pass user-defined functions in".

Doing 
```
params = lm.Parameters()
params._asteval.symtable["myfun"] = myfun
```

would work, for example.   

Probably the right thing to do is add a `usersyms` argument to `Parameters()` that does this.  Then we could deprecate the `asteval` argument.

**Comment by zobristnicholas** [comment]
That solution fixes my current problem. Thanks!

I agree that adding a usersyms argument to Parameters is cleaner than trying to modify __deepcopy__. 

It looks like this came up because asteval.Interpreter() adds all callables in usersyms to the no_deepcopy list on instantiation. So, it doesn't appear in _asteval.user_defined_symbols(). The easiest way around that is to just add the functions to the symtable after instantiation manually as you suggested.

**Comment by newville** [comment]
should be fixed in master branch with #510
